### ISSUE TITLE
Refresh tokens problems

### ISSUE THREAD

**Issue by SelaOAL** [Autodesk-Forge/forge-api-nodejs-client#7]
Hi,

I'm trying to use `oAuth2ThreeLegged.refreshToken` and  `oAuth2TwoLegged.refreshToken` to refresh the tokens but there seems to be problems with that.

`oAuth2TwoLegged` doesn't even find a `refreshToken` function.

And `oAuth2ThreeLegged` manages to refresh once, but afterwards I get:

>(node:5936) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): [object Object]

Is there a limit on update requests per minute? If so how long should I wait before updating?

**Comment by gregra81** [comment]
@SelaO-AL: 2 legged flow doesn't support refresh token. Regarding the error, can you post here the code you're running?

**Comment by SelaO-AL** [comment]
```
var ForgeSDK = require('forge-apis');

var CLIENT_ID = 'id' , CLIENT_SECRET = 'sec', REDIRECT_URL = 'http://localhost:3000';
var autoRefresh = true;

var oAuth2TwoLegged = new ForgeSDK.AuthClientTwoLegged(CLIENT_ID, CLIENT_SECRET, [
    'data:read',
    'data:write'
], autoRefresh);

var oAuth2ThreeLegged = new ForgeSDK.AuthClientThreeLegged(CLIENT_ID, CLIENT_SECRET, REDIRECT_URL, [
    'data:read',
    'data:write'
], autoRefresh);

var str2leg = 'cred str';
var str3leg = 'another cred str';

var TwoLeggedCredentials = JSON.parse(str2leg);
var ThreeLeggedCredentials = JSON.parse(str3leg);

oAuth2ThreeLegged.refreshToken(ThreeLeggedCredentials); // this will work once per credentials 
```

Entering `oAuth2ThreeLegged.refreshToken(ThreeLeggedCredentials);` again in the shell will emit a rejection.

**Comment by SelaOAL** [comment]
It might be related, but when I try to run the code from the example: [](https://github.com/Autodesk-Forge/forge-api-nodejs-client#example-api-calls)

```
BucketsApi.getBuckets({}, oAuth2TwoLegged, TwoLeggedCredentials).then(function(buckets){
    console.log('BucketsApi.getBuckets');
    console.log(buckets);
});
```

I get the same error: 

>(node:3712) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): [object Object]

**Comment by leefsmp** [comment]
The APIs that deal with user data require a 3-legged token, the one that deal with your own app data require a 2-legged. However OSS can be either 2 or 3-legged, depending if you access it from the app private storage or the user A360 account. For example if you upload a file to OSS private bucket that you created with 2-legged token, then it requires 2-legged. If you upload a file as result of creating a new item with data-management API, it requires a 3-legged token.

Hope that helps

**Comment by gregra81** [comment]
@SelaOAL  - Regarding the docs, you can actually see for every method which auth type it requires - example: https://github.com/Autodesk-Forge/forge-api-nodejs-client/blob/master/docs/HubsApi.md#authorization-2 (access_code = 3-legged, application = 2-legged)

**Comment by SelaOAL** [comment]
OK thanks.
About the error, cause it?

**Comment by gregra81** [comment]
@SelaOAL 
So, I think that what happened for you is that the promise didn't return correctly from the SDK due to errors that API returned (400 for example). I fixed that in the new version (0.2.6) - if you'll get errors, the promise will return them (inside your `catch` block). Make sure you update to the new version.

Thanks for pointing that out, I appreciate the feedback. Also, I've added this gist - https://gist.github.com/gregra81/4ac50dcac05d77aa2a495a59b658ac9b, it shows some 3-legged authentication flows.
